lib/Serialization/SerializedModuleLoader.cpp

```diff
@@ -565,17 +565,20 @@ SerializedModuleLoaderBase::findModule(AccessPathElem moduleID,
         case SearchPathKind::RuntimeLibrary: {
           isFramework = false;
 
-          bool checkTargetSpecificModule;
+          // Apple platforms always use target-specific files within a
+          // .swiftmodule directory for the stdlib; non-Apple platforms always
+          // use single-architecture swiftmodules.
+          bool checkTargetSpecificModule = false;
           if (Kind == SearchPathKind::RuntimeLibrary) {
-            // Apple platforms always use target-specific files within a
-            // .swiftmodule directory for the stdlib; non-Apple platforms
-            // always use single-architecture swiftmodules.
-            checkTargetSpecificModule = Ctx.LangOpts.Target.isOSDarwin();
+            auto result = findTargetSpecificModuleFiles();
+            if (Ctx.LangOpts.Target.isOSDarwin() || result)
+              return result;
           } else {
             auto modulePath = currPath;
             llvm::sys::path::append(modulePath, genericModuleFileName);
 
             llvm::ErrorOr<llvm::vfs::Status> statResult = fs.status(modulePath);
+
             // Even if stat fails, we can't just return the error; the path
             // we're looking for might not be "Foo.swiftmodule".
             checkTargetSpecificModule = statResult && statResult->isDirectory();
```

Comment:
I think you could simplify this logic pretty substantially:

```
          // On Apple platforms, we can assume that the runtime libraries use
          // target-specific module files within a .swiftmodule directory. This
          // was not always true on non-Apple platforms, and it's still not true
          // in other search paths, so otherwise we need to check whether
          // there's a <ModuleName>.swiftmodule directory.
          bool checkTargetSpecificModule = true;
          if (Kind != SearchPathKind::RuntimeLibrary || !Ctx.LangOpts.Target.isOSDarwin()) {
            auto modulePath = currPath;
            llvm::sys::path::append(modulePath, genericModuleFileName);

            llvm::ErrorOr<llvm::vfs::Status> statResult = fs.status(modulePath);

            // Even if stat fails, we can't just return the error; the path
            // we're looking for might not be "Foo.swiftmodule".
            checkTargetSpecificModule = statResult && statResult->isDirectory();
          }
```

Reply:
I strongly prefer this solution, I don't remember why I didn't collapse the branch.